USBDevice: Avoid forcing end device to be derived from USBDevice class

facchinm opened the pull request:
### Description

By replacing the existing callbacks with proper `mbed::callback` we can use the USBDevice infrastructure without deriving directly from USBDevice.

The patch could bring a (tiny, big ?) performance hit due to the indirection, but would make the USBDevice class usable as a singleton.

A usage example can be found here https://github.com/arduino/ArduinoCore-nRF528x-mbedos/tree/master/cores/arduino/USB and https://github.com/bcmi-labs/ArduinoCore-nRF528x-mbedos/tree/master/libraries/USBHID

PluggableUSBDevice implements USBDevice and calls phy init() just once; all device drivers (derived from mbed ones) plug on it to create a composite device but they don't derive from USBDevice.

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes


<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

facchinm committed:
Use unique include for Stream.h

maciejbocianski reviewed (changes requested):
Generally looks good, 
[tests-usb_device-basic](https://github.com/ARMmbed/mbed-os/tree/master/TESTS/usb_device/basic)  needs adaptation to your changes, you can use my code https://github.com/maciejbocianski/mbed-os/commit/092b26b355078d827af8419c49c7f3c51b96c67b

@fkjagodzinski @c1728p9 what do you think

kjbracey commented on drivers/internal/USBDevice.h:
```diff
@@ -141,21 +142,6 @@ class USBDevice: public  USBPhyEvents {
     */
     bool endpoint_add(usb_ep_t endpoint, uint32_t max_packet, usb_ep_type_t type, ep_cb_t callback = NULL);
```
If you're changing to `Callback`, could you please avoid giving it `NULL` - change to `nullptr`. Will avoid issues with some optimisation work I'm doing on `Callback`. Will save me touching this again later.

(That does make it a 5.14 change - we're not putting any C++11-isms into 5.13 patches. But this doesn't look like a patch anyway).

c1728p9 reviewed (approved):
This PR seems reasonable to me. Does the use of callback increase or decrease RAM used? Also, you'll need a few more changes if you want to make USBDevice usable without inheriting from it:
- Protected functions need to be made public so they can be called
    - Callback completion handlers such as [complete_request](https://github.com/ARMmbed/mbed-os/blob/master/drivers/internal/USBDevice.h#L401)
    - Synchronization [functions](https://github.com/ARMmbed/mbed-os/blob/master/drivers/internal/USBDevice.h#L484-L498)

Those changes could be made in a separate PR though as this change look good on its own merits.

facchinm committed:
Allow USB endpoint_add to accept mbed::Callback

In this way we can use the USBDevice infrastructure without deriving directly from USBDevice.

maciejbocianski reviewed (approved):
LGTM
